rfcs/20200705-tfx-ir.md

```diff
@@ -0,0 +1,1563 @@
+  inputs {
+    key: "transform_graph"
+    value {
+      channels {
+        producer_node_query {
+          id: "transform"
+        }
+        context_queries {...}  # The context query for the parent pipeline.
+        context_queries {...}  # The context query for the parent pipeline.
+        context_queries {...}  # The context query for the parent pipeline.
+        artifact_query {
+          type: {name: "TransformGraph"}
+        }
+        output_key: "transform_graph"
+      }
+      required: True
+    }
+  }
+  resolver_config {...}
+}
+</pre>
+
+<div align="center"><em><a id="table23">Table 23</a>: IR of the `Trainer` inside the sub-pipeline.</em></div>
+
+###### 3.3.2.4.3. Tail Barnacle
+
+The 'Tail Barnacle' will be symmetric to the 'Head Barnacle', snapshotting the
+output artifacts produced by the nodes inside the sub-pipeline so that they can
+be accessed as synchronous outputs from the outer pipeline. For simplicity, the
+represnetation of the 'Tail Barnacle' is not shown but its node id
+`tail_barnacle` will be used by the `Pusher` discussed next.
+
+###### 3.3.2.4.4. Consuming synchronous and asynchronous outputs of a sub-pipeline
+
+Consuming asynchronous outputs of a node inside a sub-pipeline is similar to the
+normal pattern in asynchronous excution pipelines. The only difference is that
+the `ContextQueries` will need to contain the context for the sub-pipeline.
+
+Consuming synchronous outputs of a node inside a sub-pipeline, on the other
+hand, will be encoded differently. As shown below, the `ProducerNodeQuery` of
+the input indicates the data model underneath that the `Pusher` is actually
+reading from the output `Channel` of the 'Tail Barnacle'.
+
+```proto
+node_info: {
+  type: {name: "Pusher"}
+  id: "pusher"
+}
+contexts {
+  contexts {...}  # parent pipeline context
+}
+inputs {
+  inputs {
+    key: "model"
+    value {
+      channels {
+        producer_node_query {
+          id: "tail_barnacle"
+        }
+        context_queries {          # The context query for the parent pipeline.
+          type {name: "pipeline"}
+          name: "my_pipeline"
+        }
+        context_queries {          # The context query for the sub-pipeline.
+          type {name: "pipeline"}
+          name: "my_sub_pipeline"
+        }
+        artifact_query {
+          type: {name: "Model"}
+        }
+        output_key: "model"
+      }
+      required: True
+    }
+  }
+  inputs {
+    key: "validation_result"
+    value {...}                    # Similar to 'model', omitted.
+  }
+  resolver_config {...}
+}
+...
+```
+
+<div align="center"><em><a id="table24">Table 24</a>: IR of `Pusher` which consumes synchronous outputs.</em></div>
+
+## 4. Future work
+
+There are three directions of future improvement on the data model and IR:
+
+1.  Better extensibility
+
+    -   `ExecutorSpec` needs to be enhanced to support more executor form
+        factors and platforms.
```

Comment:
nitpick: May be worth elaborating more on what this means (aka component environment declaration, dependency declaration, container environment declaration). I understand this is a generic phrase as it will be specified in a future RFC, but may be worthwhile tweaking language to make it more clear (had to read https://github.com/tensorflow/community/pull/271#issuecomment-664714772 to understand this one)